File: sdk/swift/Sources/TokenPayID/Internal/RemoteConfig.swift
import Foundation

/// Runtime-fetched configuration for the TOKEN PAY ID widget on Apple
/// platforms.
///
/// Mirrors the Android and JVM `RemoteConfig` classes. Integrators don't have
/// to rebuild their app to pick up new colour tokens, translated strings, or
/// feature-flag toggles — the widget calls
/// `GET /api/v1/sdk/config?client_id=…` once per launch and merges the
/// `remote` section on top of the bundled defaults.
struct TpidRemoteConfig: Sendable {
    /// `name` (or `dark.name` / `light.name`) → `#RRGGBB`.
    let themeTokens: [String: String]
    /// `lang` → `key` → translated string.
    let stringsOverride: [String: [String: String]]
    /// `name` → `true` / `false`.
    let featureFlags: [String: Bool]
    /// SDKs older than this receive a "mandatory update" banner.
    let minSdkVersion: String?
    /// SDKs older than this receive a gentle "update recommended" banner.
    let recommendedSdkVersion: String?
    /// 2.6.0 — URL the update banner's "Update" pill opens in the default
    /// browser. Populated from `remote.download_url` in `/sdk/config`.
    let downloadUrl: String?
    /// 2.6.0 — short changelog for the update banner tooltip.
    let updateNotes: String?
    /// 2.6.1 — SHA-256 of the `downloadUrl` tarball (lowercase hex).
    /// `TpidUpdater` verifies this byte-for-byte before moving the cached
    /// download into its final slot. `nil` means the server didn't emit
    /// one; the updater still succeeds but cannot self-attest integrity.
    let downloadSha256: String?

    static let empty = TpidRemoteConfig(
        themeTokens: [:], stringsOverride: [:], featureFlags: [:],
        minSdkVersion: nil, recommendedSdkVersion: nil,
        downloadUrl: nil, updateNotes: nil, downloadSha256: nil
    )

    // Process-wide cache (keyed on client id). Never throws — a network or
    // parse failure simply returns an empty config.
    private static let lock = NSLock()
    private static var cachedClientId: String?
    private static var cached: TpidRemoteConfig = .empty

    /// Fetch the config for [clientId] via [api]. Safe to call multiple times:
    /// the result is memoised for the lifetime of the process.
    static func fetch(api: ApiClient, clientId: String) async -> TpidRemoteConfig {
        lock.lock()
        if cachedClientId == clientId {
            let hit = cached
            lock.unlock()
            return hit
        }
        lock.unlock()

        let parsed: TpidRemoteConfig
        do {
            let payload = try await api.fetchSdkConfig(clientId: clientId)
            parsed = parse(payload: payload)
        } catch {
            parsed = .empty
        }

        lock.lock()
        cached = parsed
        cachedClientId = clientId
        lock.unlock()
        return parsed
    }

    /// 2.6.1 — Drop the memoised snapshot. `TpidUpdater` calls this before
    /// its force-refresh so the next `fetch(api:clientId:)` goes to the
    /// network and any hot-pushed strings / theme / flags apply live in
    /// the running widget.
    static func invalidateCache() {
        lock.lock()
        cachedClientId = nil
        cached = .empty
        lock.unlock()
    }

    private static func parse(payload: [String: Any]) -> TpidRemoteConfig {
        guard let remote = payload["remote"] as? [String: Any] else { return .empty }

        let theme: [String: String] = (remote["theme_tokens"] as? [String: Any])?
            .compactMapValues { ($0 as? String) } ?? [:]

        let strings: [String: [String: String]]
        if let rawStrings = remote["strings_override"] as? [String: Any] {
            var built: [String: [String: String]] = [:]
            for (lang, entry) in rawStrings {
                if let dict = entry as? [String: Any] {
                    let ov = dict.compactMapValues { $0 as? String }
                    if !ov.isEmpty { built[lang] = ov }
                }
            }
            strings = built
        } else {
            strings = [:]
        }

        let flags: [String: Bool]
        if let rawFlags = remote["feature_flags"] as? [String: Any] {
            // Backend emits strings like "true" / "false" — accept both.
            flags = rawFlags.compactMapValues { raw in
                if let b = raw as? Bool { return b }
                if let s = raw as? String { return s.lowercased() == "true" }
                return nil
            }
        } else {
            flags = [:]
        }

        return TpidRemoteConfig(
            themeTokens: theme,
            stringsOverride: strings,
            featureFlags: flags,
            minSdkVersion: remote["min_sdk_version"] as? String,
            recommendedSdkVersion: remote["recommended_sdk_version"] as? String,
            downloadUrl: remote["download_url"] as? String,
            updateNotes: remote["update_notes"] as? String,
            downloadSha256: remote["download_sha256"] as? String,
        )
    }

    /// Returns `true` when the currently running SDK build is older than
    /// [other]. A missing [other] is treated as "no newer build available",
    /// so the caller never shows a banner spuriously.
    func isOlderThan(_ other: String?) -> Bool {
        guard let other else { return false }
        return Self.compareSemver(TpidVersion.string, other) < 0
    }

    /// Good-enough semver comparison for the two version strings we handle.
    /// Pre-release suffixes (`-pre.N`, `-rc.N`) sort *before* the same bare
    /// version, which matches integrator intuition.
    static func compareSemver(_ a: String, _ b: String) -> Int {
        func split(_ v: String) -> (nums: [Int], hasSuffix: Bool, suffix: String) {
            let parts = v.split(separator: "-", maxSplits: 1, omittingEmptySubsequences: false)
            let core = String(parts[0])
            let suffix = parts.count > 1 ? String(parts[1]) : ""
            let nums = core.split(separator: ".").compactMap { Int($0) }
            return (nums, !suffix.isEmpty, suffix)
        }
        let lhs = split(a)
        let rhs = split(b)
        let count = max(lhs.nums.count, rhs.nums.count)
        for i in 0..<count {
            let l = i < lhs.nums.count ? lhs.nums[i] : 0
            let r = i < rhs.nums.count ? rhs.nums[i] : 0
            if l != r { return l < r ? -1 : 1 }
        }
        if lhs.hasSuffix != rhs.hasSuffix {
            // No suffix ("2.5.0") is newer than "2.5.0-pre.2".
            return lhs.hasSuffix ? -1 : 1
        }
        if lhs.suffix != rhs.suffix {
            return lhs.suffix < rhs.suffix ? -1 : 1
        }
        return 0
    }
}

File: sdk/jvm-desktop/tokenpay-id-jvm/src/main/kotlin/space/tokenpay/id/jvm/internal/RemoteConfig.kt
package space.tokenpay.id.jvm.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.concurrent.atomic.AtomicReference

/**
 * Runtime-fetched configuration for the TOKEN PAY ID widget.
 *
 * Enterprises using the native SDK can receive UI changes — new colour tokens,
 * translated strings, feature-flag toggles — without rebuilding their app.
 * The SDK asks `GET /api/v1/sdk/config?client_id=…&sdk=jvm&v=<version>` once
 * per widget launch; the server's `remote` section is applied on top of the
 * locally-bundled defaults so updates are additive.
 *
 * Contract (fields under `remote`):
 *
 *  - `theme_tokens`  : map of colour / radius / spacing overrides — applied
 *                      to the Compose `ColorScheme` when rendering the widget.
 *  - `strings_override[<lang>][<key>]` : per-language string overrides — fed
 *                      to [Strings] before the widget renders.
 *  - `feature_flags` : key/bool pairs — used to toggle alt-login methods.
 *  - `min_sdk_version` / `recommended_sdk_version` : the widget shows a
 *                      "please update" banner if the running SDK is older
 *                      than the minimum.
 *
 * Missing fields are simply ignored — the SDK falls back to its defaults.
 */
internal data class RemoteConfig(
    val themeTokens: Map<String, String> = emptyMap(),
    val stringsOverride: Map<String, Map<String, String>> = emptyMap(),
    val featureFlags: Map<String, Boolean> = emptyMap(),
    val minSdkVersion: String? = null,
    val recommendedSdkVersion: String? = null,
    /**
     * Host-facing download URL for [recommendedSdkVersion]. The widget's
     * "Update" button opens this URL in the default browser so the
     * integrator can grab the new tarball without leaving the app.
     * Defaults to an empty string, which the widget interprets as "no
     * download link available, point the user at /sdk".
     */
    val downloadUrl: String = "",
    /** Optional short changelog shown in the update banner's tooltip. */
    val updateNotes: String = "",
    /**
     * 2.6.1 — SHA-256 of the [downloadUrl] tarball, hex-encoded, lowercase.
     * Populated from `remote.download_sha256` (per-platform) in `/sdk/config`
     * and verified byte-for-byte by [TpidUpdater] before the cached file is
     * moved into place. Empty string means "server didn't provide one" —
     * the updater still succeeds in that case but cannot self-attest.
     */
    val downloadSha256: String = "",
    /** Enterprise name from the root `branding` object, if the client is active. */
    val clientName: String? = null,
) {
    companion object {
        /** In-memory cache keyed on `clientId`. One request per widget launch. */
        private val cache = AtomicReference<Pair<String, RemoteConfig>?>(null)

        /**
         * Fetch the remote config for [clientId] via [ApiClient].
         *
         * Never throws — a network failure just returns an empty config so
         * the widget still works offline. A previously-cached value survives
         * for the lifetime of the process.
         */
        suspend fun fetch(apiClient: ApiClient, clientId: String): RemoteConfig = withContext(Dispatchers.IO) {
            cache.get()?.let { if (it.first == clientId) return@withContext it.second }
            val parsed = runCatching { parse(apiClient.fetchSdkConfig(clientId)) }.getOrDefault(RemoteConfig())
            cache.set(clientId to parsed)
            parsed
        }

        /**
         * 2.6.1 — Drop the memoised value. Called by [TpidUpdater] before a
         * forced refresh so the next [fetch] actually hits the network and
         * any `remote.strings_override` / `theme_tokens` edits the operator
         * made take effect in the running widget.
         */
        fun invalidateCache() { cache.set(null) }

        private fun parse(root: JsonObject): RemoteConfig {
            val remote = (root["remote"] as? JsonObject) ?: return RemoteConfig()
            val theme = (remote["theme_tokens"] as? JsonObject)?.mapValues { (_, v) ->
                runCatching { v.jsonPrimitive.content }.getOrDefault("")
            }?.filterValues { it.isNotBlank() } ?: emptyMap()
            val stringsRoot = remote["strings_override"] as? JsonObject
            val strings: Map<String, Map<String, String>> = stringsRoot?.mapValues { (_, langEntry) ->
                (langEntry as? JsonObject)?.mapValues { (_, v) ->
                    runCatching { v.jsonPrimitive.content }.getOrDefault("")
                }?.filterValues { it.isNotBlank() } ?: emptyMap()
            }?.filterValues { it.isNotEmpty() } ?: emptyMap()
            val flags = (remote["feature_flags"] as? JsonObject)?.mapValues { (_, v) ->
                runCatching { v.jsonPrimitive.content.toBooleanStrictOrNull() ?: false }.getOrDefault(false)
            } ?: emptyMap()
            val minV = runCatching { remote["min_sdk_version"]?.jsonPrimitive?.content }.getOrNull()
            val recV = runCatching { remote["recommended_sdk_version"]?.jsonPrimitive?.content }.getOrNull()
            val dlUrl = runCatching { remote["download_url"]?.jsonPrimitive?.content }.getOrNull().orEmpty()
            val notes = runCatching { remote["update_notes"]?.jsonPrimitive?.content }.getOrNull().orEmpty()
            val sha = runCatching { remote["download_sha256"]?.jsonPrimitive?.content }.getOrNull().orEmpty()
            val clientName = ((root["branding"] as? JsonObject)?.get("client_name"))
                ?.let { runCatching { it.jsonPrimitive.content }.getOrNull() }
                ?.takeIf { it.isNotBlank() }
            return RemoteConfig(theme, strings, flags, minV, recV, dlUrl, notes, sha, clientName)
        }
    }

    /**
     * Semver comparison just good enough for the two fields we care about.
     * Pre-release suffixes (`-pre.X`, `-rc.Y`) are treated as *older* than
     * the same version without a suffix, matching user expectations.
     */
    fun isOlderThan(other: String?): Boolean {
        val otherV = other ?: return false
        val self = SdkBuildInfo.version
        return compareSemver(self, otherV) < 0
    }

    private fun compareSemver(a: String, b: String): Int {
        // Split the version into numeric core + optional pre-release suffix.
        // Using positional destructuring here crashed on clean semver like
        // "2.5.0" (single-element list) — see 2.5.0-pre.3 changelog.
        fun norm(v: String): Triple<List<Int>, Boolean, String> {
            val parts = v.split('-', limit = 2)
            val core = parts.getOrNull(0).orEmpty()
            val pre  = parts.getOrNull(1).orEmpty()
            val nums = core.split('.').mapNotNull { it.toIntOrNull() }
            return Triple(nums, pre.isBlank(), pre)
        }
        val (an, ac, ap) = norm(a)
        val (bn, bc, bp) = norm(b)
        val size = maxOf(an.size, bn.size)
        for (i in 0 until size) {
            val av = an.getOrElse(i) { 0 }
            val bv = bn.getOrElse(i) { 0 }
            if (av != bv) return av.compareTo(bv)
        }
        if (ac != bc) return if (ac) 1 else -1 // no-suffix > suffix
        return ap.compareTo(bp)
    }
}

/** Compile-time constants describing this build of the SDK. */
internal object SdkBuildInfo {
    const val version: String = "3.0.1"
    const val platform: String = "jvm"
    const val userAgent: String = "TokenPayID-JVM/$version"
    const val sdkHeader: String = "jvm-desktop/$version"
}
